app/assets/javascripts/discourse/app/controllers/topic-bulk-actions.js

```diff
@@ -101,23 +103,70 @@ export default Controller.extend(ModalFunctionality, {
   },
 
   perform(operation) {
+    this.set("processedTopicCount", 0);
+    this.send("changeBulkTemplate", "modal/bulk-progress");
     this.set("loading", true);
 
-    const topics = this.get("model.topics");
-    return Topic.bulkOperation(topics, operation)
-      .then((result) => {
-        this.set("loading", false);
-        if (result && result.topic_ids) {
-          return result.topic_ids.map((t) => topics.findBy("id", t));
-        }
-        return result;
-      })
+    return this._processChunks(operation)
       .catch(() => {
         bootbox.alert(I18n.t("generic_error"));
+      })
+      .finally(() => {
         this.set("loading", false);
       });
   },
 
+  _generateTopicChunks(allTopics) {
+    let startIndex = 0;
+    const chunkSize = 30;
+    const chunks = [];
+
+    while (startIndex < allTopics.length) {
+      let topics = allTopics.slice(startIndex, startIndex + chunkSize);
+      chunks.push(topics);
+      startIndex += chunkSize;
+    }
+
+    return chunks;
+  },
+
+  _processChunks(operation) {
+    const allTopics = this.get("model.topics");
+    const topicChunks = this._generateTopicChunks(allTopics);
+    const topicIds = [];
+
+    const tasks = topicChunks.map((topics) => () => {
```

Comment:
I've never seen this syntax before, is it a mistake?  I would expect `(topics) => {`